src/darwin/Framework/CHIP/MTRDevice_Concrete.mm

```diff
@@ -131,6 +131,67 @@ - (void)_deviceInternalStateChanged:(MTRDevice *)device;
 
 } // anonymous namespace
 
+#pragma mark - MTRDeviceMatterCPPObjectsHolder
+
+// Class to hold C++ objects that can only be manipulated on the Matter queue
+@interface MTRDeviceMatterCPPObjectsHolder : NSObject
+@property (nonatomic, readonly) ReadClient * readClient;
+@property (nonatomic, readonly) SubscriptionCallback * subscriptionCallback; // valid when and only when currentReadClient is valid
+- (void)setReadClient:(ReadClient * _Nullable)readClient subscriptionCallback:(SubscriptionCallback * _Nullable)subscriptionCallback;
+- (void)clearReadClientAndDeleteSubscriptionCallback;
+@end
+
+@implementation MTRDeviceMatterCPPObjectsHolder
+@synthesize readClient = _readClient;
+@synthesize subscriptionCallback = _subscriptionCallback;
+- (ReadClient *)readClient
+{
+    assertChipStackLockedByCurrentThread();
+    @synchronized(self) {
+        return _readClient;
+    }
+}
+- (SubscriptionCallback *)subscriptionCallback
+{
+    assertChipStackLockedByCurrentThread();
+    @synchronized(self) {
+        return _subscriptionCallback;
+    }
+}
+- (void)setReadClient:(ReadClient * _Nullable)readClient subscriptionCallback:(SubscriptionCallback * _Nullable)subscriptionCallback
+{
+    assertChipStackLockedByCurrentThread();
+    @synchronized(self) {
+        // Sanity check and log if readClient and subscriptionCallback aren't both valid or both null
+        if (((readClient == nullptr) && (subscriptionCallback != nullptr)) || ((readClient != nullptr) && (subscriptionCallback == nullptr))) {
+            MTR_LOG_ERROR("%@: setReadClient:subscriptionCallback: readClient and subscriptionCallback must both be valid or both be null %p %p", self, readClient, subscriptionCallback);
+        }
+
+        // Sanity check and log if overriding existing values
+        if ((readClient != nullptr) && (_readClient != nullptr)) {
+            MTR_LOG_ERROR("%@: setReadClient:subscriptionCallback: readClient set when current value not null %p %p", self, readClient, _readClient);
+        }
+        if (((subscriptionCallback != nullptr)) && (_subscriptionCallback != nullptr)) {
+            MTR_LOG_ERROR("%@: setReadClient:subscriptionCallback: readClient and subscriptionCallback must both be valid or both be null %p %p", self, readClient, subscriptionCallback);
+        }
+
+        _readClient = readClient;
+        _subscriptionCallback = subscriptionCallback;
+    }
+}
+- (void)clearReadClientAndDeleteSubscriptionCallback
+{
+    assertChipStackLockedByCurrentThread();
+    @synchronized(self) {
+        _readClient = nullptr;
```

Comment:
Might be worth documenting why this is done that way (i.e. the subscription callback owns and deletes the readClient)?

Reply:
See https://github.com/project-chip/connectedhomeip/issues/37420